Amazon.QLDB.Driver/result/AsyncResult.cs

```diff
@@ -16,23 +16,141 @@ namespace Amazon.QLDB.Driver
         {
-            throw new NotImplementedException();
+            if (this.isRetrieved)
+            {
+                throw new InvalidOperationException();
+            }
+
+            this.isRetrieved = true;
+            return this.ionEnumerator;
         }
 
+        /// <summary>
+        /// Gets the current query statistics for the number of read IO requests. The statistics are stateful.
+        /// </summary>
+        ///
+        /// <returns>The current IOUsage statistics.</returns>
         public IOUsage? GetConsumedIOs()
         {
-            throw new NotImplementedException();
+            return this.ionEnumerator.GetConsumedIOs();
         }
 
+        /// <summary>
+        /// Gets the current query statistics for server-side processing time. The statistics are stateful.
+        /// </summary>
+        ///
+        /// <returns>The current TimingInformation statistics.</returns>
         public TimingInformation? GetTimingInformation()
         {
-            throw new NotImplementedException();
+            return this.ionEnumerator.GetTimingInformation();
+        }
+
+        /// <summary>
+        /// Object which allows for asynchronous iteration over the individual Ion values that make up the whole result of a statement
+        /// execution against QLDB.
+        /// </summary>
+        private class IonAsyncEnumerator : BaseIonEnumerator, IAsyncEnumerator<IIonValue>
+        {
+            private readonly Session session;
+            private readonly string txnId;
+            private new IEnumerator<ValueHolder> currentEnumerator;
+            private string nextPageToken;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="IonAsyncEnumerator"/> class.
+            /// </summary>
+            ///
+            /// <param name="session">The parent session that represents the communication channel to QLDB.</param>
+            /// <param name="txnId">The unique ID of the transaction.</param>
+            /// <param name="statementResult">The result of the statement execution.</param>
+            internal IonAsyncEnumerator(Session session, string txnId, ExecuteStatementResult statementResult)
+            {
+                this.session = session;
+                this.txnId = txnId;
+                this.currentEnumerator = statementResult.FirstPage.Values.GetEnumerator();
+                this.nextPageToken = statementResult.FirstPage.NextPageToken;
+
+                if (statementResult.ConsumedIOs != null)
+                {
+                    this.readIOs = statementResult.ConsumedIOs.ReadIOs;
+                    this.writeIOs = statementResult.ConsumedIOs.WriteIOs;
+                }
+
+                if (statementResult.TimingInformation != null)
+                {
+                    this.processingTimeMilliseconds = statementResult.TimingInformation.ProcessingTimeMilliseconds;
+                }
```

Comment:
Isn't the constructor logic shared, so it can be in baseIonEnumerator?

Reply:
Shared in BaseIonEnumerator.